express-joi-validation.js

```diff
@@ -87,11 +89,57 @@ module.exports = function generateJoiMiddlewareInstance (cfg) {
         } else {
           res
             .status(opts.statusCode || cfg.statusCode || 400)
-            .end(buildErrorString(ret, type));
+            .end(buildErrorString(ret, `request ${type}`));
         }
       };
     };
   });
 
   return instance;
+
+  function response(schema, _options) {
+    const options = _options || {}
+    const {
+      joi: opts,
+      validCode = 200,
+      failedCode = 500,
+      respondOnFail = true,
+      throwOnFail = false
+    } = options
+    return validator
+
+    function validator(req, res, next) {
+      res.sendValidJson = sendValidJson
```

Comment:
I've done some thinking on this PR. I would rather not add a new method to `res`, but instead wrap the existing `res.json` because:

* Avoids pollution of `res` object
* Developers could easily forget to use `sendValidJson` instead of `json`
* Removes need for `validCode` param since you can instead use the regular `res.status(NUMBER)` and `res.json(DATA)` methods
* I'd rather that the module follows the existing express patterns of `send`, `json`, `end` etc. and not return a promise
* It would follow the general design of this module as being something that works (mostly) transparently of the actual request handler code

So something like the code below where `throwOnFail` and `respondOnFail ` are not required since we pass to `next` similar to other API functions. Also `failedCode` is changed to `statusCode` to match the other API functions.

```js
function response (schema, _options) {
  // would need to verify safety of approach, since if res.json uses "this" internally
  // we'd need to do "const json = res.json.bind(res)" or similar right?
  const resJson = res.json

  res.json = function (json) {
    const ret = Joi.validate(json, schema, opts)

    if (ret.error) {
      if (opts.passError || cfg.passError) {
          ret.type = 'response'
          next(e)
        } else {
          res
            .status(opts.statusCode || cfg.statusCode || 500)
            .end(buildErrorString(...args...))
        }
    } else {
      // return res.json result to retain express compatibility
      return resJson(ret.value)
    }
  }
}
```

I may be overlooking some reasoning for decisions made, so let me know if so.

Reply:
Ping @maikelmclauflin

Reply:
@evanshortiss Yeah, I tried this way. Express ends up going in a circular fashion and res.json gets called by res.send, which gets called by res.json. There's probably a way around it, i just found that it was way easier to think about (from a dev / first iteration perspective) if different method + user can opt out if they'd like
But I'll see if I can figure out a way around

Reply:
@maikelmclauflin that's a bummer. It seems to work for me on latest express using the code below. As for opt-out, I think that can be done by simply not adding the middleware?

```js
const express = require('express')

const app = express()

const mw = (req, res, next) => {
    const originalJson = res.json.bind(res)
    res.json = function (data) {
        console.log('sending json', data) // this will print if the mw works
        originalJson(data)
    }
    next()
}

app.get('/', mw, (req, res) => {
    res.json({ hello: 'world'})
})

app.listen(3030, (err) => {
    if (err) {
        throw err
    }
    console.log('listening on 3030')
})
```

Reply:
odd. i'll try again. maybe i was using send.